src/replication.c

```diff
@@ -0,0 +1,1067 @@
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <sys/uio.h>
+#include <netdb.h>
+#include <unistd.h>
+#include <sys/epoll.h>
+#include <fcntl.h>
+#include "replication.h"
+#include "istgt_integration.h"
+#include "replication_misc.h"
+#include "istgt_crc32c.h"
+#include "istgt_misc.h"
+cstor_conn_ops_t cstor_ops = {
+	.conn_listen = replication_listen,
+	.conn_connect = replication_connect,
+};
+
+#define build_replica_io_hdr() \
+{\
+	rio = (zvol_io_hdr_t *)malloc(sizeof(zvol_io_hdr_t));\
+	rio->opcode = cmd->opcode;\
+	rio->io_seq = cmd->io_seq;\
+	rio->offset = cmd->offset;\
+	rio->len    = cmd->data_len;\
+}
+#define build_replica_mgmt_hdr() \
+{\
+	rmgmtio = (zvol_io_hdr_t *)malloc(sizeof(zvol_io_hdr_t));\
+	rmgmtio->opcode = mgmt_opcode;\
+	rmgmtio->len    = data_len;\
+}
+
+/*
+#define build_mgmt_cmd() \
+{\
+	mgmt_cmd = malloc(sizeof(mgmt_cmd_t));\
+	mgmt_cmd->opcode = opcode;\
+	mgmt_cmd->data_len = malloc \
+}
+*/
+
+#define build_rcmd() \
+{\
+	rcmd = malloc(sizeof(rcmd_t));\
+	rcmd->opcode = cmd->opcode;\
+	rcmd->offset = cmd->offset;\
+	rcmd->data_len = cmd->data_len;\
+	rcmd->io_seq = cmd->io_seq;\
+	rcmd->ack_recvd = false;\
+	rcmd->rcommq_ptr = cmd;\
+	rcmd->status = 0;\
+	rcmd->iovcnt = cmd->iovcnt;\
+	for (i=1; i < rcmd->iovcnt + 1; i++) {\
+		rcmd->iov[i].iov_base = cmd->iov[i].iov_base;\
+		rcmd->iov[i].iov_len = cmd->iov[i].iov_len;\
+	}\
+}
+
+#define check_read_error() {\
+	if (count < 0)\
+	{\
+		if (errno != EAGAIN)\
+		{\
+			REPLICA_ERRLOG("read error");\
+			close (events[i].data.fd);\
+			break;\
+		} else {\
+			continue;\
+		}\
+	} else if(count == 0) {\
+		close (events[i].data.fd);\
+		break;\
+	}\
+}
+
+#define check_for_blockage() {\
+	uint64_t current_lba = cmd->offset; \
+	uint64_t current_lbE = current_lba + cmd->data_len; \
+	uint64_t pending_lba = pending_cmd->offset; \
+	uint64_t pending_lbE = pending_lba + pending_cmd->data_len; \
+	if ((current_lbE < pending_lba)  || \
+			(pending_lbE < current_lba)) { \
+		cmd_blocked = false; \
+	}\
+}
+
+int
+send_mgmtio(int fd, zvol_op_code_t mgmt_opcode, void *buf, size_t data_len) {
+	zvol_io_hdr_t *rmgmtio = NULL;
+	int iovcnt, i, nbytes = 0;
+	int rc = 0;
+	struct iovec iovec[2];
+	build_replica_mgmt_hdr();
+	iovec[0].iov_base = rmgmtio;
+	nbytes = iovec[0].iov_len = sizeof(zvol_io_hdr_t);
+	iovec[1].iov_base = buf;
+	iovec[1].iov_len = rmgmtio->len;
+	nbytes += rmgmtio->len;
+	iovcnt = 2;
+	while (nbytes) {
+		rc = writev(fd, iovec, 2);//Review iovec in this line
+		if (rc < 0) {
```

Comment:
need to handle EAGAIN return value, and at multiple other places where writev is used

Reply:
EPOLLOUT event of fd seems to be not part of epoll. This should be good.